### PR TITLE
Support git-extract without remote repo

### PR DESCRIPTION
**Pull request by ricmatsui**
Including echo if true helper and global has remote

### CODE REVIEW

File: src/git-extract [hunk 1]
```diff
@@ -51,20 +55,18 @@ function preconditions {
 
 
 function steps {
-  if [ "$initial_open_changes" = true ]; then
-    echo "stash_open_changes"
-  fi
+  local has_remote="$(has_remote_url)"
```
Comment (charlierudolph) [inline src/git-extract]: how about setting a global `$has_remote` in `src/helpers/helpers.sh`? Especially with the coming changes it seems like that would be useful,

File: features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature [hunk 2]
```diff
@@ -0,0 +1,101 @@
+Feature: git extract: resolving conflicts between main branch and extracted commits (with open changes and without remote origin)
+
+  As a developer extracting a commit that conflicts with the main branch and without a remote repository
+  I want to be given the choice to resolve the conflicts or abort
+  So that I can finish the operation as planned or postpone it to a better time.
```
Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature]: This shouldn't be necessary but should instead just reference another user value. Probably
```
(see ../with_remote_origin/with_open_changes.feature)
```

File: features/git-extract/cherry_pick_conflict/without_remote_origin/without_open_changes.feature [hunk 3]
```diff
@@ -0,0 +1,85 @@
+Feature: git extract: resolving conflicts between main branch and extracted commits (without open changes or remote repo)
+
+  (see ./with_open_changes.feature)
```
Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/without_remote_origin/without_open_changes.feature]: (see ../with_remote_origin/with_open_changes.feature)

File: features/git-extract/multiple_commits/without_remote_origin/with_open_changes.feature [hunk 4]
```diff
@@ -0,0 +1,40 @@
+Feature: git extract: extracting multiple commits (with open changes and without remote repo)
+
+  As a developer working on a feature branch with many commits around an unrelated issue and without a remote repository
+  I want to be able to extract all of these commits into their own branch
+  So that the issue can be reviewed separately and my feature branch remains focussed.
```
Comment (charlierudolph) [inline features/git-extract/multiple_commits/without_remote_origin/with_open_changes.feature]: (see ../with_remote_origin/with_open_changes.feature)

File: features/git-extract/multiple_commits/without_remote_origin/without_open_changes.feature [hunk 5]
```diff
@@ -0,0 +1,34 @@
+Feature: git extract: extracting multiple commits (without open changes or remote repo)
+
+  (see ./with_open_changes.feature)
```
Comment (charlierudolph) [inline features/git-extract/multiple_commits/without_remote_origin/without_open_changes.feature]: (see ../with_remote_origin/with_open_changes.feature)

File: features/git-extract/single_commit/without_remote_origin/with_open_changes.feature [hunk 6]
```diff
@@ -0,0 +1,35 @@
+Feature: git extract: extracting a single commit (with open changes and without remote origin)
+
+  (see ../multiple_commits/with_open_changes.feature)
```
Comment (charlierudolph) [inline features/git-extract/single_commit/without_remote_origin/with_open_changes.feature]: (see ../multiple_commits/with_remote_origin/with_open_changes.feature)

File: features/git-extract/single_commit/without_remote_origin/without_open_changes.feature [hunk 7]
```diff
@@ -0,0 +1,31 @@
+Feature: git extract: extracting a single commit (without open changes or remote repo)
+
+  (see ../multiple_commits/with_open_changes.feature)
```
Comment (charlierudolph) [inline features/git-extract/single_commit/without_remote_origin/without_open_changes.feature]: (see ../multiple_commits/with_open_changes.feature)

File: features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature (hunk 2 above)
Comment (ricmatsui) [inline features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature]: No problem, was just emphasizing the ` and without a remote repository`

**Comment by charlierudolph** [comment]
Looks great to me. @kevgo @allewun any thoughts?

File: features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      | BRANCH   | COMMAND                 |
+      | refactor | git cherry-pick --abort |
+      | refactor | git checkout main       |
+      | main     | git branch -d refactor  |
+      | main     | git checkout feature    |
+      | feature  | git stash pop           |
+    And I end up on the "feature" branch
+    And I again have an uncommitted file with name: "uncommitted" and content: "stuff"
+    And there is no "refactor" branch
+    And I have the following commits
```
Comment (kevgo) [inline features/git-extract/cherry_pick_conflict/without_remote_origin/with_open_changes.feature]: I have recently added a step called [And I am left with my original commits](https://github.com/Originate/git-town/blob/master/features/step_definitions/commit_steps.rb#L21). I'm in the process of cleaning up the feature specs and applying it everywhere. Since the commit table here is exactly the same as the one that defines the commits in the `Background` part, you can use it here to DRY up this spec even further.

File: features/git-extract/cherry_pick_conflict/without_remote_origin/without_open_changes.feature [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    When I run `git extract --abort`
+    Then it runs the Git commands
+      | BRANCH   | COMMAND                 |
+      | refactor | git cherry-pick --abort |
+      | refactor | git checkout main       |
+      | main     | git branch -d refactor  |
+      | main     | git checkout feature    |
+    And I end up on the "feature" branch
+    And there is no "refactor" branch
+    And I have the following commits
```
Comment (kevgo) [inline features/git-extract/cherry_pick_conflict/without_remote_origin/without_open_changes.feature]: Use `And I am left with my original commits` here as well

File: features/git-extract/single_commit/without_remote_origin/with_open_changes.feature [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+      | refactor | git stash pop                                |
+    And I end up on the "refactor" branch
+    And I still have an uncommitted file with name: "uncommitted" and content: "stuff"
+    And I have the following commits
+      | BRANCH   | LOCATION | MESSAGE         | FILE NAME     |
+      | main     | local    | main commit     | main_file     |
+      | feature  | local    | feature commit  | feature_file  |
+      |          |          | refactor commit | refactor_file |
+      | refactor | local    | main commit     | main_file     |
+      |          |          | refactor commit | refactor_file |
```
Comment (kevgo) [inline features/git-extract/single_commit/without_remote_origin/with_open_changes.feature]: Technically running `git extract --undo` should undo the changes, i.e. remove the created branch. Its likely that this already works, given that our infrastructure knows how to undo individual operations. You could add the feature spec for this here. Given the size of this PR, this is probably done best in a separate PR.

File: features/git-extract/single_commit/without_remote_origin/without_open_changes.feature [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+      | main     | git checkout -b refactor main                |
+      | refactor | git cherry-pick <%= sha 'refactor commit' %> |
+    And I end up on the "refactor" branch
+    And I have the following commits
+      | BRANCH   | LOCATION | MESSAGE         | FILE NAME     |
+      | main     | local    | main commit     | main_file     |
+      | feature  | local    | feature commit  | feature_file  |
+      |          |          | refactor commit | refactor_file |
+      | refactor | local    | main commit     | main_file     |
+      |          |          | refactor commit | refactor_file |
```
Comment (kevgo) [inline features/git-extract/single_commit/without_remote_origin/without_open_changes.feature]: Same here, specify the behavior of `git extract --undo` here.

File: src/helpers/helpers.sh [hunk 12]
```diff
@@ -38,4 +38,5 @@ source "$current_dir/configuration.sh" "$@"
 if [ "$(is_git_repository)" == true ]; then
   export initial_branch_name=$(get_current_branch_name)
   export initial_open_changes=$(has_open_changes)
+  export has_remote=$(has_remote_url)
```
Comment (kevgo) [inline src/helpers/helpers.sh]: That's a cool idea.

**Comment by kevgo** [comment]
Very cool change @ricmatsui, and again very well done! :ship: from my end as well once the remaining comments are addressed. 

Btw, given your great and high quality contributions, and the fact that you totally get what we are doing here, I have given you committer status. So you can now push into our repo, and can use our CI server. Please never add anything directly into master, but always do pull requests, and wait for at least one LGTM or :ship: from another committer before shipping. Please also always use `git ship`, i.e. do squash merges when merging into master. Besides that, Git Town is your oyster now!

Feel free to `git ship` this branch yourself, and welcome to the Git Town team, Ricardo!  :tada:

**Comment by ricmatsui** [comment]
Sounds good, the comments have been addressed, and I'll make another PR for the new items.

Thanks! Definitely looking forward to contributing more! Yeah, I think this is a great tool to make Git so much more productive and cleaner. Maybe I'll get a few :rocket:s too!
